indico/modules/events/layout/models/menu.py

```diff
@@ -292,12 +282,16 @@ class MenuEntry(MenuEntryMixin, db.Model):
         nullable=False,
         default=False
     )
-    #: Which group of people should the menu entry be viewable by
-    access = db.Column(
-        PyIntEnum(MenuEntryAccess),
-        nullable=False,
-        default=MenuEntryAccess.everyone,
+    #: ACLs for page visibility
+    acl_entries = db.relationship(
+        'MenuEntryPrincipal',
+        backref='menu_entry',
+        cascade='all, delete-orphan',
+        collection_class=set
     )
+    # TODO not really sure what I need this for
+    acl = association_proxy('acl_entries', 'principal', creator=lambda v: MenuEntryPrincipal(principal=v))
+
```

Comment:
I'm really struggling to do the right thing here, maybe you can help. I copied this off how `RoomPrincipal` and the likes are doing it, however when I try to save values it will tell me that the key `menu_entries` is not found on `ClassManager of <class 'indico.modules.events.registration.models.forms.RegistrationForm'>` for example when adding a registration form to the permissions.

Am I setting the backrefs incorrectly? Do I need to add something to User/Group/RegistrationForm/etc to make this work? Or is this a red herring and I'm doing something else wrong?

Reply:
I cannot reproduce that error with the current code in the PR. In fact editing works when I replace `acl_entries` with `acl` in the form class.

However, that field is a bit broken for objects not yet in the DB (e.g. when creating a new page). You may want to consider using the field we also use for attachments which is a bit simpler ("protected" switch instead of the "protected/inheriting" toggle):

```patch
diff --git a/indico/modules/events/layout/controllers/menu.py b/indico/modules/events/layout/controllers/menu.py
index 2945bc093e..9240f44e74 100644
--- a/indico/modules/events/layout/controllers/menu.py
+++ b/indico/modules/events/layout/controllers/menu.py
@@ -90,7 +90,7 @@ class RHMenuEntryEditBase(RHMenuBase):

 class RHMenuEntryEdit(RHMenuEntryEditBase):
     def _process(self):
-        defaults = FormDefaults(self.entry)
+        defaults = FormDefaults(self.entry, protected=self.entry.is_self_protected)
         form_kwargs = {}
         if self.entry.is_user_link:
             form_cls = MenuLinkForm
diff --git a/indico/modules/events/layout/forms.py b/indico/modules/events/layout/forms.py
index e75f675c9e..8dfd0b2e4b 100644
--- a/indico/modules/events/layout/forms.py
+++ b/indico/modules/events/layout/forms.py
@@ -15,8 +15,9 @@ from indico.modules.events.layout import theme_settings
 from indico.modules.events.layout.util import get_css_file_data, get_logo_data, get_plugin_conference_themes
 from indico.modules.users import NameFormat
 from indico.util.i18n import _, orig_string
-from indico.web.forms.base import IndicoForm
-from indico.web.forms.fields import EditableFileField, FileField, IndicoEnumSelectField, IndicoProtectionField
+from indico.web.forms.base import IndicoForm, generated_data
+from indico.web.forms.fields import (AccessControlListField, EditableFileField, FileField, IndicoEnumSelectField,
+                                     IndicoProtectionField)
 from indico.web.forms.fields.enums import IndicoEnumRadioField
 from indico.web.forms.fields.principals import PrincipalListField
 from indico.web.forms.validators import HiddenUnless, UsedIf
@@ -163,12 +164,14 @@ class MenuUserEntryFormBase(IndicoForm):
     title = StringField(_('Title'), [DataRequired()])
     is_enabled = BooleanField(_('Show'), widget=SwitchWidget())
     new_tab = BooleanField(_('Open in a new tab'), widget=SwitchWidget())
-    acl_entries = PrincipalListField(_('Visibility'), event=lambda form: form.event, allow_groups=True, allow_event_roles=True, allow_category_roles=True, allow_registration_forms=True, allow_speakers=True)
-    protection_mode = IndicoProtectionField(_('Protection mode'), protected_object=lambda form: form.protected_object)
+    # acl_entries = PrincipalListField(_('ACL'), event=lambda form: form.event, allow_groups=True,
+    #                                  allow_event_roles=True, allow_category_roles=True, allow_registration_forms=True,
+    #                                  allow_speakers=True)
+    # protection_mode = IndicoProtectionField(_('Protection mode'), protected_object=lambda form: form.protected_object)

     def __init__(self, *args, event=None, **kwargs):
         self.event = event
-        self.protected_object = kwargs.get('entry', None)
+        # self.protected_object = kwargs.get('entry', None)
         super().__init__(*args, **kwargs)


@@ -179,10 +182,21 @@ class MenuLinkForm(MenuUserEntryFormBase):
 class MenuPageForm(MenuUserEntryFormBase):
     html = TextAreaField(_('Content'), [DataRequired()], widget=CKEditorWidget(images=True, html_embed=True))

+    protected = BooleanField(_('Protected'), widget=SwitchWidget())
+    acl = AccessControlListField(_('Access control list'), [UsedIf(lambda form, field: form.protected.data)],
+                                 allow_groups=True, allow_external_users=True, allow_event_roles=True,
+                                 allow_category_roles=True, allow_registration_forms=True,
+                                 event=lambda form: form.event,
+                                 description=_('The list of users and groups allowed to access the page'))
+
     def __init__(self, *args, ckeditor_upload_url, **kwargs):
         self.ckeditor_upload_url = ckeditor_upload_url
         super().__init__(*args, **kwargs)

+    @generated_data
+    def protection_mode(self):
+        return ProtectionMode.protected if self.protected.data else ProtectionMode.inheriting
+

 class AddImagesForm(IndicoForm):
     image = FileField('Image', multiple_files=True, accepted_file_types='image/jpeg,image/jpg,image/png,image/gif')
```


---


however, i played around a bit and this would work with the other widget. but as you can see there's much more complexity...

```patch
diff --git a/indico/modules/events/layout/forms.py b/indico/modules/events/layout/forms.py
index e75f675c9e..7cc4e5167e 100644
--- a/indico/modules/events/layout/forms.py
+++ b/indico/modules/events/layout/forms.py
@@ -12,6 +12,7 @@ from wtforms.validators import DataRequired, Optional, ValidationError
 from indico.core.config import config
 from indico.core.db.sqlalchemy.protection import ProtectionMode
 from indico.modules.events.layout import theme_settings
+from indico.modules.events.layout.models.menu import MenuEntry
 from indico.modules.events.layout.util import get_css_file_data, get_logo_data, get_plugin_conference_themes
 from indico.modules.users import NameFormat
 from indico.util.i18n import _, orig_string
@@ -163,13 +164,6 @@ class MenuUserEntryFormBase(IndicoForm):
     title = StringField(_('Title'), [DataRequired()])
     is_enabled = BooleanField(_('Show'), widget=SwitchWidget())
     new_tab = BooleanField(_('Open in a new tab'), widget=SwitchWidget())
-    acl_entries = PrincipalListField(_('Visibility'), event=lambda form: form.event, allow_groups=True, allow_event_roles=True, allow_category_roles=True, allow_registration_forms=True, allow_speakers=True)
-    protection_mode = IndicoProtectionField(_('Protection mode'), protected_object=lambda form: form.protected_object)
-
-    def __init__(self, *args, event=None, **kwargs):
-        self.event = event
-        self.protected_object = kwargs.get('entry', None)
-        super().__init__(*args, **kwargs)
 
 
 class MenuLinkForm(MenuUserEntryFormBase):
@@ -177,10 +171,15 @@ class MenuLinkForm(MenuUserEntryFormBase):
 
 
 class MenuPageForm(MenuUserEntryFormBase):
+    acl = PrincipalListField(_('Visibility'), event=lambda form: form.event, allow_groups=True, allow_event_roles=True,
+                             allow_category_roles=True, allow_registration_forms=True, allow_speakers=True)
+    protection_mode = IndicoProtectionField(_('Protection mode'), protected_object=lambda form: form.protected_object)
     html = TextAreaField(_('Content'), [DataRequired()], widget=CKEditorWidget(images=True, html_embed=True))
 
-    def __init__(self, *args, ckeditor_upload_url, **kwargs):
+    def __init__(self, *args, ckeditor_upload_url, event, **kwargs):
         self.ckeditor_upload_url = ckeditor_upload_url
+        self.event = event
+        self.protected_object = kwargs.get('entry', MenuEntry(event=event))
         super().__init__(*args, **kwargs)
 
 
diff --git a/indico/modules/events/layout/models/menu.py b/indico/modules/events/layout/models/menu.py
index e6e05b504f..e917c4398a 100644
--- a/indico/modules/events/layout/models/menu.py
+++ b/indico/modules/events/layout/models/menu.py
@@ -43,7 +43,7 @@ class MenuEntryType(RichIntEnum):
 # TODO move protection mixin here?
 class MenuEntryMixin:
     def __init__(self, **kwargs):
-        event = kwargs.pop('event', kwargs.get('event'))
+        event = kwargs.pop('event', None)
         super().__init__(**kwargs)
         # XXX: not calling this `event` since this one should NOT use
         # the relationship to avoid mixing data from different DB sessions
@@ -317,10 +317,6 @@ class MenuEntry(MenuEntryMixin, ProtectionMixin, db.Model):
         nullable=False
     )
 
-
-    
-
-
     #: The Event containing the menu entry
     event = db.relationship(
         'Event',
@@ -418,10 +414,7 @@ class MenuEntry(MenuEntryMixin, ProtectionMixin, db.Model):
 
     @property
     def protection_parent(self):
-        if self.parent_id is None:
-            return self.event
-        else:
-            return self.parent
+        return self.parent or getattr(self, '_event_ref', self.event)
 
 
 class EventPage(db.Model):
diff --git a/indico/web/forms/fields/protection.py b/indico/web/forms/fields/protection.py
index 411f8425ba..d857f8d42d 100644
--- a/indico/web/forms/fields/protection.py
+++ b/indico/web/forms/fields/protection.py
@@ -26,9 +26,10 @@ class IndicoProtectionField(IndicoEnumRadioField):
         self.can_inherit_protection = self.protected_object.protection_parent is not None
         self.is_unlisted_event = isinstance(self.protected_object, db.m.Event) and self.protected_object.is_unlisted
         # TODO is this just legacy and needs to be merged?
-        #if not self.can_inherit_protection and not self.is_unlisted_event:
-        #    kwargs['skip'] = {ProtectionMode.inheriting}
-        super().__init__(*args, enum=ProtectionMode, skip=self.protected_object.disallowed_protection_modes, **kwargs)
+        skip = set(self.protected_object.disallowed_protection_modes)
+        if not self.can_inherit_protection and not self.is_unlisted_event:
+            skip.add(ProtectionMode.inheriting)
+        super().__init__(*args, enum=ProtectionMode, skip=skip, **kwargs)
 
     def render_protection_message(self):
         protected_object = self.get_form().protected_object
```

Reply:
Thank you, applying your patch it magically works for me now as well! What I like about the full protection field is that it has explicit information on how it is accessible when inheriting is selected, I think that is worth the extra complexity.